File: server/src/plugins/runtime/python/proxy/file_serve.py
from __future__ import annotations

import os
from contextlib import suppress
from typing import TypedDict

from _camera_ui_tools.camera_ui_rpc import CloseHandler, RPCClient, RPCMethod
from plugins.runtime.python.namespaces import NamespaceManager


class FileServeStat(TypedDict):
    exists: bool
    size: int


class PluginFileServer:
    def __init__(self, proxy: RPCClient, plugin_id: str) -> None:
        self.__proxy = proxy
        self.__plugin_id = plugin_id
        self.__close_handler: CloseHandler | None = None

    async def register(self) -> None:
        subject = NamespaceManager.plugin_file_serve_rpc(self.__plugin_id)
        self.__close_handler = await self.__proxy.register_handler(subject, self)

    async def close(self) -> None:
        if self.__close_handler:
            await self.__close_handler()
            self.__close_handler = None

    @RPCMethod
    async def statFile(self, file_path: str) -> FileServeStat:
        try:
            if not os.path.isfile(file_path):
                return {"exists": False, "size": 0}
            return {"exists": True, "size": os.path.getsize(file_path)}
        except OSError:
            return {"exists": False, "size": 0}

    @RPCMethod
    async def readFileChunk(self, file_path: str, offset: int, length: int) -> bytes:
        with open(file_path, "rb") as f:
            f.seek(offset)
            return f.read(length)

    @RPCMethod
    async def deleteFile(self, file_path: str) -> None:
        with suppress(OSError):
            os.remove(file_path)

File: server/src/plugins/runtime/node/proxy/fileServe.ts
import { close, open, read, stat, unlink } from 'node:fs';
import { promisify } from 'node:util';

import { RPCClass, RPCMethod } from '@camera.ui/rpc';

import { NamespaceManager } from '../../../../rpc/namespaces.js';

import type { RPCClient } from '@camera.ui/rpc';
import type { FileServeStat, PluginFileServeInterface } from '../../../../rpc/interfaces/download.js';

const statAsync = promisify(stat);
const openAsync = promisify(open);
const readAsync = promisify(read);
const closeAsync = promisify(close);
const unlinkAsync = promisify(unlink);

@RPCClass
export class PluginFileServer implements PluginFileServeInterface {
  #closeHandler?: () => Promise<void>;

  constructor(
    private readonly proxy: RPCClient,
    private readonly pluginId: string,
  ) {}

  public async register(): Promise<void> {
    const subject = NamespaceManager.pluginFileServeRpc(this.pluginId);
    this.#closeHandler = await this.proxy.registerHandler(subject, this, { withoutDecorators: true });
  }

  public async close(): Promise<void> {
    await this.#closeHandler?.();
    this.#closeHandler = undefined;
  }

  @RPCMethod
  public async statFile(filePath: string): Promise<FileServeStat> {
    try {
      const stats = await statAsync(filePath);
      return { exists: stats.isFile(), size: stats.size };
    } catch {
      return { exists: false, size: 0 };
    }
  }

  @RPCMethod
  public async readFileChunk(filePath: string, offset: number, length: number): Promise<Uint8Array> {
    const fd = await openAsync(filePath, 'r');
    try {
      const buffer = Buffer.allocUnsafe(length);
      const { bytesRead } = await readAsync(fd, buffer, 0, length, offset);
      return buffer.subarray(0, bytesRead);
    } finally {
      await closeAsync(fd);
    }
  }

  @RPCMethod
  public async deleteFile(filePath: string): Promise<void> {
    await unlinkAsync(filePath).catch(() => {});
  }
}
